src/Workspaces/CSharp/Portable/LanguageServices/CSharpTypeInferenceService.TypeInferrer.cs

```diff
@@ -176,16 +176,20 @@ private IEnumerable<TypeInferenceInfo> GetTypesSimple(ExpressionSyntax expressio
 
             private IEnumerable<TypeInferenceInfo> InferTypeInArrowExpressionClause(ArrowExpressionClauseSyntax arrowClause)
             {
-                switch (arrowClause.Parent)
-                {
-                    case PropertyDeclarationSyntax propertyDeclaration:
-                        return InferTypeInPropertyDeclaration(propertyDeclaration);
-                    case BaseMethodDeclarationSyntax baseMethodDeclaration:
-                        return InferTypeInBaseMethodDeclaration(baseMethodDeclaration);
-                    case AccessorDeclarationSyntax accessorDeclaration:
-                        return InferTypeInAccessorDeclaration(accessorDeclaration);
+                var parentSymbol = SemanticModel.GetDeclaredSymbol(arrowClause.Parent);
+
+                switch (parentSymbol)
+                {
+                    case IPropertySymbol propertySymbol:
+                        return propertySymbol.Type != null
+                            ? SpecializedCollections.SingletonEnumerable(new TypeInferenceInfo(propertySymbol.Type))
+                            : SpecializedCollections.EmptyEnumerable<TypeInferenceInfo>();
+                    case IMethodSymbol methodSymbol:
+                        return methodSymbol.ReturnType != null 
+                            ? SpecializedCollections.SingletonEnumerable(new TypeInferenceInfo(methodSymbol.ReturnType))
+                            : SpecializedCollections.EmptyEnumerable<TypeInferenceInfo>();
                     default:
-                        return SpecializedCollections.EmptyEnumerable<TypeInferenceInfo>();
+                        return SpecializedCollections.EmptyEnumerable<TypeInferenceInfo>(); 
```

Comment:
I would just extract a method to get the type from the symbol:
```c#
private static ITypeSymbol GetSymbolReturnType(ISymbol symbol)
{
    switch (symbol)
    {
        case IPropertySymbol property: return property.Type;
        case IMethodSymbol method: return method.ReturnType;
        default: return null;
    }
}
```
which you could use and then only do the `type != null` check once here and avoid all the duplication.

Reply:
but you don't have to listen to me anymore :smile: i'm just giving suggestions

Reply:
Note that this last change should fix local functions too, but we still need a test for that.

Reply:
Well it's another good suggestion, so I did that. I also added unit tests for local functions, where type inference now also works as expected.